### PR TITLE
[DPE-1974][DPE-2387][DPE-2563] -- Large deployments config

### PR DESCRIPTION
**Pull request by Mehdi-Bendriss**
## Issue
This PR implements user provided options in the context of the large deployment implementation. This PR is a prequel and pre-condition to the peer cluster relations implementation - which might explain some references to the not-yet implemented peer_cluster_relation in the code.
  
Namely this PR implements:
- [DPE-1974](https://warthogs.atlassian.net/browse/DPE-1974): user provided roles
   - this include temperature attributes of the nodes
   - validation of the roles + quorum
   - switch from auto-generated to user-provided roles and vice-versa 
- [DPE-2563](https://warthogs.atlassian.net/browse/DPE-2563): cluster_name config options
   - can only be set on deployment - ignored afterwards
   - can be inherited from peer cluster relation
- [DPE-2387](https://warthogs.atlassian.net/browse/DPE-2387): init_hold onfig options
   - can only be set on deployment, ignored afterwards
- Introducing pydantic for all data models / relation data
- [DPE-2732](https://warthogs.atlassian.net/browse/DPE-2732): Integration tests
- [DPE-2387](https://warthogs.atlassian.net/browse/DPE-2387): Unit tests

[DPE-1974]: https://warthogs.atlassian.net/browse/DPE-1974?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-2563]: https://warthogs.atlassian.net/browse/DPE-2563?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-2387]: https://warthogs.atlassian.net/browse/DPE-2387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-2732]: https://warthogs.atlassian.net/browse/DPE-2732?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-2387]: https://warthogs.atlassian.net/browse/DPE-2387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+
+class OpenSearchProvidedRolesException(OpenSearchError):
+    """Exception class for events when the user provided node roles will violate quorum."""
+
+
+class OpenSearchPeerClustersManager:
+    """This class covers the configuration changes depending on certain actions."""
+
+    def __init__(self, charm):
```
Comment (deusebio) [inline lib/charms/opensearch/v0/opensearch_peer_clusters.py:55]: **(strong) suggestion** I would in general avoid to provide the charm as argument to this. I understand this is fairly common, but this really creates a bit of circular  dependency here, if we were to type this, since the Charm needs to import `OpenSearchPeerClustersManager`, that however would need to import the charm type. 

I suppose eventually this class should have knowledge of:
1. Config of the charm, represented by the pydantic object
2. Intra-application peer relation (since needs to write/read the `DeploymentDescrition`)
3. Extra-application relation (since the PeerClusterRelation will eventually be implemented)

I noticed that `_opensearch` is only used to have the `is_node_up` method. Maybe instead of passing this object, can just be easier to provide either a boolean or inject a lambda function that does that query, but probably the entire object is really not needed, and created a bit of tricky dependency tree

I understand that the other classes are implemented in this way, but (if you agree with these points) I'd rather start addressing this when creating new classes, than addressing this in a single refactoring, becuase that refactoring would be huge and probably never feasible. I would attach this bit by bit instead...

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            # if validation called on new unit: it means it will start and maintain the quorum
+            #    (called on the latest unit to be configured and brought online)
+            # if validation called on existing cluster we should expect an odd number in the sum
+            return
+
+        raise OpenSearchProvidedRolesException(PClusterWrongNodesCountForQuorum)
+
+    def is_peer_cluster_relation_set(self):
+        """Return whether the peer cluster relation is established."""
+        return PeerClusterRelationName in self._charm.model.relations
```
Comment (deusebio) [inline lib/charms/opensearch/v0/opensearch_peer_clusters.py:313]: Maybe I would also factor this to a different PR, when this relation will be implemeted. If one wants to still keep this function, one can probably `return False`, with a comment that this will be implemented once the relation exists

**Review by deusebio — COMMENTED** [review]
Great job Mehdi! The code looks fairly all right. I have few suggestions, comments, but nothing really major or critical. Beside few nipticks and comments, possibly the only two things worth mentioning here are

1. You are using a bit pydantic class to represents configurations as well as relation data. I would be inclined to nudge you to use and implement a similar structure as the one we did in kafka, postgresql, mysql, etc using the `data_models` data platform lib. This can very well be done in a separate PR, but I believe it definitely worth addressing to keep charms consistent across our products. 
2. I was a bit confused where the cross-application relation was, since part of the code refers to this, but then realized that this was not yet implemented. I believe it makes sense to centralize all these modifications into the next PR, just for the sake of making also that PR a bit clearer (without the need to have important parts in this PR instead). So my suggestion would be to take out the bits around the relation (e.g. I have added some comments) 

I have not checked closely the more OpenSearch business logic, but I trust the other reviewers judgement on this. Also, some of the comments/questions may also come from my limited understanding of how the multi-cloud deployment would play out effectively. Please bear with me! ;)

**Review by marcoppenheimer — CHANGES_REQUESTED** [review]
Left some suggestions. Generally it's hard to review as there are a LOT of code-jumps with the object structure so it's hard to be sure. Happy to approve once comments are addressed though, as I imagine refactoring that aspect would be prohibitively time-consuming.

Taking a higher level view, I think that more effort could be invested on making the actual 'design' of large deployments more parseable between events. 

I'll try to summarise my understanding of the design, then leave thoughts (assuming I'm not wildly off base in my grokking):

- We have 3 config options for an application:
    - `init-hold` - freezes start until joined with another app
    - `cluster-name` - key to match apps together
    - `roles` - useful descriptor for app function
- During deployment, if `role=cluster_manager`, build a JSON representation of the cluster topology
- During config-changed + relation-changed, if `role=cluster_manager`, build a JSON representation of the cluster topology
- If not `role=cluster_manager`, respond to `relation-changed` events


If I understood that correctly, then I think the class layout could be a bit more expressive, as the current implementation of `OpenSearchPeerClustersManager` is VERY complex.

As `init-hold` is involved with `start` and general set-up events, it's probably best to separate this from the `OpenSearchPeerClustersManager` class, and instead have a dedicated `ClustersCoordinator` (name whatever) object that handles things like 'can I start', 'do I need to update stuff' etc, that all return booleans. These defensive checks can be called directly from `charm.py` before doing things. e.g `if OpenSearchCluster.can_start: restart.emit()`. For example, in the ZK charms we have `ZooKeeperCluster`, which also acts like a coordinator for cluster management - https://github.com/canonical/zookeeper-operator/blob/main/src/cluster.py, and we call things from the charm directly - https://github.com/canonical/zookeeper-operator/blob/5e9a0530c55cf1c5a37b5308ad977133aab6ed22/src/charm.py#L215-L232

For `roles`, I think it might make sense to have a dedicated `Roles` class, that has some attrs `config`/`relation_data` (pulled from the charm attr), where it can calculate any roles diffs, things that should be done for different roles etc. I guess here would be a bunch of business logic about what we should be doing with each role + role-change.

Then, lastly, I think we should have some `MultiClusterRelation` class, that contains a representation of `deployment_desc`, and all the `cluster_manager` operations like `run()` and `*_cluster_setup()`, because 'setting up' the cluster is (I think...) just a matter of setting relation data. 


tl;dr - I think the abstractions makes it hard to understand what's going on. Focus on expressing core Charming concepts of 'starting', 'config validation' and 'relation-data getting/setting', APPLIED to the context of Large Clusters, rather than hiding all that logic inside it. Not that I'm suggesting custom events, but if there were these events, what would they be? What behaviors do we want charm authors to respond to? 'you_are_good_to_start', 'role_changed_update_stuff', 'config_borked' etc.

File: lib/charms/opensearch/v0/models.py [hunk 3]
```diff
@@ -16,13 +20,56 @@
...
+class Model(ABC, BaseModel):
+    """Base model class."""
+
+    def to_str(self) -> str:
+        """Deserialize object into a string."""
+        return self.json()
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Deserialize object into a dict."""
+        return self.dict()
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/models.py:32]: I did it temporarily to hide the pydantic methods, as the APIs seem to be different in Pydantic `1.x` vs `2.x`.  I could not use the v2 because of some rust dependency issues - but the plan would be to move away from 1.x and then get rid of this

File: lib/charms/opensearch/v0/models.py [hunk 4]
```diff
@@ -16,13 +20,56 @@
...
+        """Create a new instance of this class from a json/dict repr."""
+        return cls(**input_dict)
+
+    @classmethod
+    def from_str(cls, input_str_dict: str):
+        """Create a new instance of this class from a stringified json/dict repr."""
+        return cls.from_dict(json.loads(input_str_dict))
+
+    def __eq__(self, other) -> bool:
+        """Implement equality."""
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/models.py:45]: It has to do with the lists comparison indeed. Unfortunately not all lists can be represented as sets, i.e: lists with duplicate or non-hashable items.

File: lib/charms/opensearch/v0/models.py [hunk 5]
```diff
@@ -40,23 +87,118 @@ def is_data(self):
...
+
+    cluster_name: str
+    init_hold: bool
+    roles: List[str]
+    data_temperature: Optional[str] = None
+
+    @root_validator
+    def set_node_temperature(cls, values):  # noqa: N805
+        """Set and validate the node temperature."""
+        allowed_temps = ["hot", "warm", "cold", "frozen"]
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/models.py:171]: I'm not sure how to do that with pydantic, it seems we can have it as a `Field(.., const=True)` but then this would be (de)serialized.. Otherwise a plain `@property` but then it probably wouldn't be worth having it.. 
Thoughts?

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+class OpenSearchPeerClustersManager:
+    """This class covers the configuration changes depending on certain actions."""
+
+    def __init__(self, charm):
+        self._charm = charm
+        self._opensearch = charm.opensearch
+        self.peer_cluster_data = RelationDataStore(self._charm, PeerClusterRelationName)
+
+    def run(self) -> None:
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py:60]: Your assumption is correct. Currently, the only entrypoint is `config_changed` which happens either during the bootstrapping of the charm or of a config change event. But the idea would be to also call it on `peer_cluster_relation_*` events

File: lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_relation_peer_cluster.py:1]: Yes, that will be filled in the next PR

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    """Exception class for events when the user provided node roles will violate quorum."""
+
+
+class OpenSearchPeerClustersManager:
+    """This class covers the configuration changes depending on certain actions."""
+
+    def __init__(self, charm):
+        self._charm = charm
+        self._opensearch = charm.opensearch
+        self.peer_cluster_data = RelationDataStore(self._charm, PeerClusterRelationName)
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py:58]: same reply as [here](https://github.com/canonical/opensearch-operator/pull/133#discussion_r1366721892)

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py (hunk 2 above)
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py:313]: same reply as [here](https://github.com/canonical/opensearch-operator/pull/133#discussion_r1366721892)

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 8]
```diff
@@ -265,6 +308,7 @@ def _on_peer_relation_changed(self, event: RelationChangedEvent):
         ):
             # we defer because we want the temporary status to be updated
             event.defer()
+            self.on.update_status.emit()
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:311]: This was to ensure the deferred event will be replayed (as opposed to waiting for the potential next event in the queue) - this is no longer applicable, as we adopted a plain dummy [`DeferEventTrigger`](https://github.com/canonical/opensearch-operator/blob/main/lib/charms/opensearch/v0/opensearch_base_charm.py#L113C39-L113C56)

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 9]
```diff
@@ -410,6 +452,15 @@ def _on_config_changed(self, _: ConfigChangedEvent):
...
                 self.model.get_relation(PeerRelationName)
             )
+
+        if self.unit.is_leader():
+            # run peer cluster manager processing
+            self.opensearch_peer_cm.run()
+        elif not self.opensearch_peer_cm.deployment_desc():
+            # deployment desc not initialized yet by leader
+            event.defer()
+            return
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:462]: You're right - the `run` must only be done from the leader for now - and I considered having the check on leader in `run` but I wanted the distinction of behavior (`run` on leader VS `wait if not ran` on non leaders) to be clear when reading the code of the charm (which is the usual reading entrypoint)

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 10]
```diff
@@ -519,9 +570,9 @@ def _start_opensearch(self, event: EventBase) -> None:  # noqa: C901
         if self.opensearch.is_started():
             try:
                 self._post_start_init()
-                self.status.clear(WaitingToStart)
             except (OpenSearchHttpError, OpenSearchNotFullyReadyError):
                 event.defer()
+                self.on.update_status.emit()
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:575]: same reply as [here](https://github.com/canonical/opensearch-operator/pull/133#discussion_r1366746321)

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 11]
```diff
@@ -577,6 +636,7 @@ def _start_opensearch(self, event: EventBase) -> None:  # noqa: C901
             self.peers_data.delete(Scope.UNIT, "starting")
             self.unit.status = BlockedStatus(ServiceStartError)
             event.defer()
+            self.on.update_status.emit()
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:639]: same reply as [here](https://github.com/canonical/opensearch-operator/pull/133#discussion_r1366746321)

**Review by delgod — APPROVED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: If you replay an event immediately, it will have the same relation-data context during execution as the event that `emitted` it. 
What are you expecting to have changed? Is it some API call?

**Review by delgod — DISMISSED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 12]
```diff
@@ -202,12 +212,45 @@ def _on_start(self, event: StartEvent):
             event.defer()
             return
 
+        self.status.clear(TLSNotFullyConfigured)
+
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:225]: I don't see why it shouldn't be possible, but I also don't see a valid reason for it I think. If I understand correctly, you're essentially postponing the whole validation / configuration of the cluster to the `start_opensearch` instead of NOT emitting a start unless it qualified as `startable`, correct? in which case we still would defer in non-leader units in case `deployment_desc is None`. Did I miss something? 

*For the last bullet point, non-leaders do not start from `*-changed event`, they start from a deferred `_on_start`

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/opensearch_base_charm.py]: Yes, it's `self.health.apply()` - which essentially fetches (through the OpenSearch REST API) the health of the cluster

File: lib/charms/opensearch/v0/helper_cluster.py [hunk 13]
```diff
@@ -58,6 +58,11 @@ def recompute_nodes_conf(nodes: List[Node]) -> Dict[str, Node]:
         if updated_node:
             nodes_by_name[updated_node.name] = updated_node
 
+        # this is in case the nodes had had a change in the "base roles", i.e: removed "ingest"
```
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply lib/charms/opensearch/v0/helper_cluster.py]: Very good catch, thank you! The logic there was wrong: this part of the function essentially handles the case where we are moving from: `user provided roles => automatically generated` 
This logic worked because I tested on "single cluster" deployments, so any node should have contained the `cluster_manager` role. But In a multi cluster deployment this is wrong, I implemented a `refill_node_with_default_roles` function that takes into consideration the multi cluster constraint where the current cluster may not have had `cluster_manager`

**Comment by Mehdi-Bendriss** [comment]
Thanks @marcoppenheimer for the review ! 
I will try to give more context on the implementation. (Though for a much deep view, it would be best to check the spec DA019) 

-----
The logic is slightly complex, as the structure is meant to support in the same way:
- simple classical deployments 
- large (multi-cluster) deployments - which is perceived in the same way as a simple deployment from within the charm class.
- switch from user-provided config options (i.e: roles) to charm-generated ones, and vice-versa
-----

As for the config options, they are pretty tightly coupled with each others (the exhaustive algorithm can be found in the DA019 spec):
- `cluster_name` - crucial as it's the key to nodes forming 1 cluster. It can be: 
    - auto-generated - then immutable
    - set by the user on deployment - then immutable
    - inherited by the peer cluster relation - then immutable (to be implemented in the next PR)
- `init_hold` - whether a cluster should start either:
    - right after deployment (default mode, i.e: simple clusters) - `False`
    - only after being related to and configured by another peer-cluster - `True` 
- `roles` - the roles the nodes of a cluster should have - a role is a simple string, but they are not simply "descriptive". Each role maps to a defined function the node should perform in a cluster, or an underlined infrastructure config. They can be:
    - auto-generated: default mode, the charm generates the best possible combination of roles while maintaining the quorum ==> validation not needed as the charm has control over the logic 
        - They are computed then broadcasted by the leader unit in various events i.e (`peer_rel_changed`, `peer_rel_departed` etc...)
    - user provided: validation is a must, as this is tightly coupled with `init_hold`, for ex: `init_hold == False  and  roles == [data, ml] ===> invalid` because it's missing the `cluster_manager` role. etc.. (for the whole logic check the DA019 spec)
    - there can be a switch from auto-generated, to user-provided and vice-versa, in which case again validation and/or re-computing / broadcasting of roles must be done in a cluster, then a rolling restart.

-----
The flow is: 
1. First deployment
2. `config-changed`
    1. leader unit builds a `DeploymentDescription` of the current cluster. This object will contain a few things but the most crucial:
        1. **descriptive items** (i.e: the StartMode of the cluster (generated roles or user-provided), the deployment_type (main cluster_manager, failover or others), cluster_name, init_hold, roles etc..)
        2. **a reporting** of the deployment state
        3. **a list of pending directives**: set of actions for each node to perform 
    2. non-leader units wait until this description is built 
3. `start`:
   1. if the `DeploymentDescription` is built, a combination of its content is what defines whether the current cluster `can_start`
4. In case of a subsequent change by the user for ex of the config option, we're back to `#2` above
5. `DeploymentDescription` is built / updated, and broadcasted in the peer relation data
6. `peer-relation-changed` is triggered in each unit 
7. each unit reconfigures itself and rolling restart

-----
The reason there is a strong reliance on the `*-relation-changed` event, is that opensearch nodes are self-contained and autonomous, they're the ones responsible for configuring themselves and joining a cluster. So in order to perform a change on each node, the process is to rely on the relation data to broadcast a new config and have all the nodes pick it up and apply it on themselves, to then rolling restart.

-----

If I understood correctly, I think what you see as `ClustersCoordinator` is effectively what I refer to as `OpenSearchPeerClustersManager`:

- though it refers to things like `can_start`, it doesn't go as low as the daemon / technical level, what it does is essentially computes whether a node is `allowed` to start as per :
    1. the combination of the config options
    2. the validity of the roles 
    3. etc...
- works on the "current" cluster level - not across multiple clusters  

(The `OpenSearchPeerClusterRelation` is the subject of the next PR.)

**Review by marcoppenheimer — DISMISSED** [review]

File: lib/charms/opensearch/v0/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply lib/charms/opensearch/v0/models.py:31]: Oh, I see. In this case it is fine I suppose!

File: lib/charms/opensearch/v0/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply lib/charms/opensearch/v0/models.py:44]: All right, that's fine if you really say that you need a custom comparison

File: lib/charms/opensearch/v0/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply lib/charms/opensearch/v0/models.py]: Well, I just meant that it is a bit better to do the following:

```
class TestClass(BaseModel):
     _VALUES=["a", "b"]
     a: str
     @root_validator
     def _validate(cls, values):
         if values["a"] not in cls._VALUES:
             raise TypeError()
         return values
```

Or something along this line

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py:64]: Yes, all right. This was probably due to the misunderstanding on the PR scope. Resolving now

File: lib/charms/opensearch/v0/opensearch_peer_clusters.py (hunk 1 above)
Comment (deusebio) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py]: Yes, but just to be clear, what suggested by Marc is just a workaround to type things, but it does not address the circular dependency issue, which - in general - makes also components way more tightly connected one another, reducing/limiting modularity, testability and separation of concerns. Preveting the typing (as addressed by the workaround) is only one of the drawbacks, but not the only one 

I'd argue circular dependency is a bit of an anti-pattern in software engineering, which I would in general encourage to avoid whenever possible

Comment (deusebio) [reply lib/charms/opensearch/v0/opensearch_peer_clusters.py]: anyhow, I don't think it is worth to stop the PR if this requires substantial refactoring outside the scope of this PR

**Review by deusebio — APPROVED** [review]

**Review by deusebio — APPROVED** [review]

**Review by marcoppenheimer — APPROVED** [review]
